psl-core/src/main/java/org/linqs/psl/reasoner/sgd/term/SGDObjectiveTerm.java

```diff
@@ -87,23 +85,73 @@ public float evaluate(float[] variableValues) {
     /**
      * Minimize the term by changing the random variables and return how much the random variables were moved by.
      */
-    public float minimize(int iteration, float[] variableValues) {
+    public float minimize(float[] variableValues, int iteration, float learningRate,
+                          Map<Integer, Float> accumulatedGradientSquares,
+                          Map<Integer, Float> accumulatedGradientMean,
+                          Map<Integer, Float> accumulatedGradientVariance,
+                          boolean adaGrad, boolean adam, boolean coordinateStep) {
         float movement = 0.0f;
+        float variableStep = 0.0f;
+        float newValue = 0.0f;
+        float partial = 0.0f;
+        float dot = dot(variableValues);
 
-        for (int i = 0 ; i < size; i++) {
-            float dot = dot(variableValues);
-            float gradient = computeGradient(i, dot);
-            float gradientStep = gradient * (learningRate / iteration);
+        for (int i = 0; i < size; i++) {
+            partial = computePartial(i, dot);
+            variableStep = computeVariableStep(variableIndexes[i], iteration, learningRate, partial,
+                    accumulatedGradientSquares, accumulatedGradientMean, accumulatedGradientVariance,
+                    adaGrad, adam);
 
-            float newValue = Math.max(0.0f, Math.min(1.0f, variableValues[variableIndexes[i]] - gradientStep));
+            newValue = Math.max(0.0f, Math.min(1.0f, variableValues[variableIndexes[i]] - variableStep));
             movement += Math.abs(newValue - variableValues[variableIndexes[i]]);
             variableValues[variableIndexes[i]] = newValue;
+
+            if (coordinateStep) {
+                dot = dot(variableValues);
+            }
         }
 
         return movement;
     }
 
-    private float computeGradient(int varId, float dot) {
+    private float computeVariableStep(int variableIndex, int iteration, float learningRate, float partial,
+                                      Map<Integer, Float> accumulatedGradientSquares,
+                                      Map<Integer, Float> accumulatedGradientMean,
+                                      Map<Integer, Float> accumulatedGradientVariance,
+                                      boolean adaGrad, boolean adam) {
+        float beta1 = 0.9f * (float)Math.pow(1.0f - 1.0e-8f, iteration - 1);
+        float beta2 = 0.999f;
+        float mean_hat = 0.0f;
+        float variance_hat = 0.0f;
+        float step = 0.0f;
+        float adaptedLearningRate = 0.0f;
+
+        if (adaGrad) {
```

Comment:
Does this want to be a switch statement? Will we have others besides adaGrad and adam?

Reply:
We may have other gradient descent based optimizers. 
Ways I see to set the optimizer are 
1: Multiple boolean options for each optimizer (Current method).
2: A single string option for the optimizer, then we can use a switch statement.
3: Create child classes of SGDReasoner and SGDObjectiveTerm which override the computeVariableStep method and then set the reasoner with the INFERENCE_REASONER option.